// @angular-react/core/esm2020/src/lib/angular-react-browser.module.mjs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { NgModule } from '@angular/core';
import { BrowserModule, ɵDomRendererFactory2 } from '@angular/platform-browser';
import { AngularReactRendererFactory } from './renderer/renderer';
import * as i0 from "@angular/core";
export class AngularReactBrowserModule {
}
/** @nocollapse */ AngularReactBrowserModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: AngularReactBrowserModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
/** @nocollapse */ AngularReactBrowserModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: AngularReactBrowserModule, imports: [BrowserModule], exports: [BrowserModule] });
/** @nocollapse */ AngularReactBrowserModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: AngularReactBrowserModule, providers: [{ provide: ɵDomRendererFactory2, useClass: AngularReactRendererFactory }], imports: [[BrowserModule], BrowserModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.1", ngImport: i0, type: AngularReactBrowserModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [BrowserModule],
                    exports: [BrowserModule],
                    providers: [{ provide: ɵDomRendererFactory2, useClass: AngularReactRendererFactory }],
                }]
        }] });
